File: tests/spec/lua/test_runtime_stream_eof_semantics.lua
local luaunit = require("luaunit")

require("kaitaistruct")
local stringstream = require("string_stream")

TestRuntimeStreamEofSemantics = {}

function TestRuntimeStreamEofSemantics:test_short_read_rewinds_seekable_stream()
    local ks = KaitaiStream(stringstream("abcd"))

    luaunit.assertErrorMsgContains("requested 5 bytes, but only 4 bytes available", function()
        ks:read_bytes(5)
    end)

    luaunit.assertEquals(ks:pos(), 0)
end

function TestRuntimeStreamEofSemantics:test_read_bytes_aligns_after_bit_reads()
    local ks = KaitaiStream(stringstream("\170\187"))

    luaunit.assertEquals(ks:read_bits_int_be(4), 0xA)
    luaunit.assertEquals(ks:read_bytes(1), "\187")
end

File: tests/spec/python/spec/test_runtime_stream_eof_semantics.py
import io
import unittest

from kaitaistruct import KaitaiStream, EndOfStreamError


class _NonSeekableBytesIO(io.BytesIO):
    def seekable(self):
        return False

    def seek(self, *args, **kwargs):
        raise io.UnsupportedOperation("not seekable")

    def tell(self):
        raise io.UnsupportedOperation("not seekable")


class TestRuntimeStreamEofSemantics(unittest.TestCase):
    def test_short_read_rewinds_seekable_stream(self):
        ks = KaitaiStream(io.BytesIO(b"abcd"))

        with self.assertRaises(EndOfStreamError):
            ks.read_bytes(5)

        self.assertEqual(ks.pos(), 0)

    def test_short_read_non_seekable_stream_has_no_rewind(self):
        ks = KaitaiStream(_NonSeekableBytesIO(b"abcd"))

        with self.assertRaises(EndOfStreamError):
            ks.read_bytes(5)

    def test_read_bytes_aligns_after_bit_reads(self):
        ks = KaitaiStream(io.BytesIO(b"\xAA\xBB"))
        self.assertEqual(ks.read_bits_int_be(4), 0xA)
        self.assertEqual(ks.read_bytes(1), b"\xBB")
